File: src/lib/openaiService.ts

interface MappingInfo {
  name: string;
  rows: Array<{
    sourceColumn: {
      malcode: string;
      table: string;
      column: string;
    };
    targetColumn: {
      malcode: string;
      table: string;
      column: string;
    };
    dataType: string;
    transformationLogic?: string;
  }>;
}

interface OpenAIResponse {
  sqlQuery: string;
  testData: any[];
  validationResults: {
    isValid: boolean;
    message: string;
    executedResults?: any[];
    errors?: string[];
    suggestions?: string[];
  };
}

export class OpenAIService {
  private apiKey: string;
  private baseUrl = 'https://api.openai.com/v1/chat/completions';

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  async generateSQL(mappingInfo: MappingInfo): Promise<string> {
    console.log('Generating SQL with OpenAI for mapping:', mappingInfo.name);
    
    const prompt = `
Given the following data mapping information, generate a SQL query that demonstrates the relationships between source and target columns:

Mapping File: ${mappingInfo.name}

Mappings:
${mappingInfo.rows.map(row => 
  `- Source: ${row.sourceColumn.malcode}.${row.sourceColumn.table}.${row.sourceColumn.column} (${row.dataType})
  - Target: ${row.targetColumn.malcode}.${row.targetColumn.table}.${row.targetColumn.column}
  - Transformation: ${row.transformationLogic || 'Direct mapping'}`
).join('\n')}

Please generate a comprehensive SQL query that:
1. Selects data from the source tables
2. Maps the columns according to the specifications
3. Includes appropriate JOINs if multiple tables are involved
4. Uses proper SQL syntax and formatting

Return only the SQL query without additional explanation.
    `;

    const response = await this.callOpenAI(prompt);
    return response.choices[0]?.message?.content || '';
  }

  async generateTestData(mappingInfo: MappingInfo, sqlQuery: string): Promise<any[]> {
    console.log('Generating test data with OpenAI for SQL validation');
    
    const prompt = `
Given the following SQL query and mapping information, generate comprehensive test data that covers various scenarios for validation:

SQL Query:
${sqlQuery}

Mapping Information:
${mappingInfo.rows.map(row => 
  `- ${row.sourceColumn.table}.${row.sourceColumn.column} -> ${row.targetColumn.table}.${row.targetColumn.column} (${row.dataType})`
).join('\n')}

Generate test data scenarios that include:
1. Happy path scenarios with normal data
2. Edge cases (null values, empty strings, boundary values)
3. Data type validation scenarios
4. Performance testing scenarios with varying data sizes
5. Special characters and internationalization cases

Return the test data as a JSON array where each object represents a test scenario with:
- test_scenario: string (description of the test case)
- All relevant columns from the mapping with appropriate test values
- expected_result: string (what should happen with this data)

Generate at least 8-10 different test scenarios. Return only valid JSON without any markdown formatting or additional text.
    `;

    const response = await this.callOpenAI(prompt);
    const content = response.choices[0]?.message?.content || '[]';
    
    try {
      return JSON.parse(content);
    } catch (error) {
      console.error('Failed to parse test data JSON:', error);
      return this.generateFallbackTestData(mappingInfo);
    }
  }

  async validateSQLWithTestData(sqlQuery: string, testData: any[]): Promise<{
    isValid: boolean;
    message: string;
    executedResults?: any[];
    errors?: string[];
    suggestions?: string[];
  }> {
    console.log('Validating SQL with OpenAI using test data');
    
    const prompt = `
Analyze the following SQL query and test data to provide validation results:

SQL Query:
${sqlQuery}

Test Data Sample:
${JSON.stringify(testData.slice(0, 3), null, 2)}

Please analyze and provide:
1. Whether the SQL query is syntactically correct
2. Whether the query would work with the provided test data structure
3. Potential performance issues or optimizations
4. Data type compatibility issues
5. Any logical errors in the query structure
6. Suggestions for improvement

Simulate executing the query with the test data and provide mock results.

Respond with a JSON object in this exact format:
{
  "isValid": boolean,
  "message": "string describing overall validation result",
  "executedResults": [array of mock query results],
  "errors": [array of error messages if any],
  "suggestions": [array of improvement suggestions]
}

Return only valid JSON without any markdown formatting or additional text.
    `;

    const response = await this.callOpenAI(prompt);
    const content = response.choices[0]?.message?.content || '{}';
    
    try {
      return JSON.parse(content);
    } catch (error) {
      console.error('Failed to parse validation JSON:', error);
      return {
        isValid: false,
        message: 'Failed to validate SQL query',
        errors: ['OpenAI response parsing error']
      };
    }
  }

  async processComplete(mappingInfo: MappingInfo): Promise<OpenAIResponse> {
    console.log('Running complete OpenAI analysis pipeline');
    
    // Step 1: Generate SQL
    const sqlQuery = await this.generateSQL(mappingInfo);
    
    // Step 2: Generate test data for the SQL
    const testData = await this.generateTestData(mappingInfo, sqlQuery);
    
    // Step 3: Validate SQL with the test data
    const validationResults = await this.validateSQLWithTestData(sqlQuery, testData);
    
    return {
      sqlQuery,
      testData,
      validationResults
    };
  }

  private async callOpenAI(prompt: string) {
    const response = await fetch(this.baseUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: 'gpt-4o-mini',
        messages: [
          {
            role: 'user',
            content: prompt
          }
        ],
        temperature: 0.7,
        max_tokens: 2000
      })
    });

    if (!response.ok) {
      throw new Error(`OpenAI API error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  private generateFallbackTestData(mappingInfo: MappingInfo): any[] {
    return [
      {
        test_scenario: "Happy Path - Normal Data",
        customer_id: 12345,
        customer_name: "John Doe",
        order_amount: 150.75,
        expected_result: "Should return valid customer order"
      },
      {
        test_scenario: "Edge Case - NULL Values",
        customer_id: 11111,
        customer_name: "Bob Wilson",
        order_amount: null,
        expected_result: "Should handle NULL values gracefully"
      }
    ];
  }
}

// Get OpenAI API key from localStorage or prompt user
export const getOpenAIKey = (): string | null => {
  return localStorage.getItem('openai_api_key');
};

export const setOpenAIKey = (key: string): void => {
  localStorage.setItem('openai_api_key', key);
};

export const createOpenAIService = (): OpenAIService | null => {
  const apiKey = getOpenAIKey();
  if (!apiKey) {
    return null;
  }
  return new OpenAIService(apiKey);
};

File: backend/openai_service.py

import json
import logging
from typing import List, Dict, Any
from fastapi import HTTPException

from config import get_openai_client, AZURE_OPENAI_DEPLOYMENT_NAME
from models import MappingInfo, ValidationResults

logger = logging.getLogger(__name__)

def call_azure_openai(messages: List[Dict[str, str]], max_tokens: int = 2000) -> str:
    """Call Azure OpenAI with the given messages"""
    client = get_openai_client()
    if not client:
        raise HTTPException(status_code=500, detail="Azure OpenAI is not configured")
    
    try:
        response = client.chat.completions.create(
            model=AZURE_OPENAI_DEPLOYMENT_NAME,
            messages=messages,
            max_tokens=max_tokens,
            temperature=0.7,
            top_p=0.9,
        )
        return response.choices[0].message.content
    except Exception as e:
        logger.error(f"Azure OpenAI API call failed: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Azure OpenAI API call failed: {str(e)}")

def generate_sql_query(mapping_info: MappingInfo) -> str:
    """Generate SQL query using Azure OpenAI based on mapping information"""
    
    mapping_details = "\n".join([
        f"Source: {row['sourceColumn']['malcode']}.{row['sourceColumn']['table']}.{row['sourceColumn']['column']} -> "
        f"Target: {row['targetColumn']['malcode']}.{row['targetColumn']['table']}.{row['targetColumn']['column']}"
        for row in mapping_info.rows
    ])

    messages = [
        {
            "role": "system",
            "content": """You are an expert SQL developer specializing in data transformation and ETL processes. 
            Generate efficient SQL queries for data mapping scenarios between source and target systems.
            Focus on:
            1. Creating proper SELECT statements with transformations
            2. Handling data type conversions appropriately
            3. Including necessary JOINs when multiple tables are involved
            4. Adding WHERE clauses for data quality checks
            5. Using appropriate SQL functions for data transformation"""
        },
        {
            "role": "user",
            "content": f"""Generate a SQL query for the following data mapping:

Mapping Name: {mapping_info.name}

Mappings:
{mapping_details}

Requirements:
- Create a SELECT query that transforms source data to target format
- Include all mapped columns
- Add appropriate data type conversions
- Include comments explaining the transformation logic
- Make the query production-ready

Please provide only the SQL query without additional explanations."""
        }
    ]
    
    return call_azure_openai(messages, max_tokens=1500)

def generate_test_data(mapping_info: MappingInfo, sql_query: str) -> List[Dict[str, Any]]:
    """Generate test data using Azure OpenAI based on mapping and SQL query"""
    
    column_info = "\n".join([
        f"{row['targetColumn']['table']}.{row['targetColumn']['column']} ({row.get('dataType', 'string')})"
        for row in mapping_info.rows
    ])

    messages = [
        {
            "role": "system",
            "content": """You are a test data generator expert. Create realistic test data that covers various scenarios including:
            1. Normal/valid data cases
            2. Edge cases (nulls, empty strings, boundary values)
            3. Data quality issues (for testing validation)
            4. Different data types and formats
            
            Always return valid JSON array format."""
        },
        {
            "role": "user",
            "content": f"""Generate test data for this mapping scenario:

Mapping: {mapping_info.name}

Target Columns:
{column_info}

SQL Query:
{sql_query}

Generate 10-15 test records that include:
- 5-7 normal valid records
- 2-3 edge case records (nulls, empty values)
- 2-3 boundary value records
- 1-2 potential data quality issue records

Return as a JSON array of objects. Each object should have keys matching the target column names.
Example format: [{"column1": "value1", "column2": "value2"}, ...]

Provide only the JSON array without additional text."""
        }
    ]
    
    response = call_azure_openai(messages, max_tokens=2000)
    
    try:
        # Try to parse the JSON response
        test_data = json.loads(response)
        if not isinstance(test_data, list):
            raise ValueError("Response is not a JSON array")
        return test_data
    except (json.JSONDecodeError, ValueError) as e:
        logger.error(f"Failed to parse test data JSON: {e}")
        # Return fallback test data
        return [
            {"id": i, "sample_column": f"test_value_{i}", "status": "valid"}
            for i in range(1, 6)
        ]

def validate_sql_query(sql_query: str, test_data: List[Dict[str, Any]]) -> ValidationResults:
    """Validate SQL query using Azure OpenAI"""
    
    test_data_sample = json.dumps(test_data[:3], indent=2) if test_data else "No test data provided"

    messages = [
        {
            "role": "system",
            "content": """You are a SQL validation expert. Analyze SQL queries for:
            1. Syntax correctness
            2. Performance optimization opportunities
            3. Data quality and integrity checks
            4. Best practices compliance
            5. Potential runtime issues
            
            Provide structured feedback with specific recommendations."""
        },
        {
            "role": "user",
            "content": f"""Validate this SQL query:

SQL Query:
{sql_query}

Sample Test Data:
{test_data_sample}

Provide validation results in this format:
- Overall Assessment: (Valid/Invalid with reasoning)
- Syntax Issues: (List any syntax problems)
- Performance Suggestions: (List optimization recommendations)
- Data Quality Concerns: (List potential data issues)
- Best Practice Recommendations: (List improvements)

Be specific and actionable in your feedback."""
        }
    ]
    
    response = call_azure_openai(messages, max_tokens=1500)
    
    # Parse the response to determine validity
    is_valid = "invalid" not in response.lower() and "error" not in response.lower()
    
    return ValidationResults(
        isValid=is_valid,
        message=response,
        executedResults=None,
        errors=[] if is_valid else ["Validation issues found - see message for details"],
        suggestions=[response] if not is_valid else []
    )
